oci/layout/oci_src_test.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/stretchr/testify/require"
+)
+
+const RemoteLayerContent = "This is the remote layer content"
+
+func TestMain(m *testing.M) {
+	httpServer, serverStopSignal, err := startRemoteLayerServer()
+	if err != nil {
+		println("Error starting test TLS server", err.Error())
+		os.Exit(1)
+	}
+
+	code := m.Run()
+
+	stopServer(httpServer, serverStopSignal)
+	os.Exit(code)
+}
+
+func TestGetBlobForRemoteLayers(t *testing.T) {
+	imageSource := createImageSource(t, &types.SystemContext{})
+	layerInfo := types.BlobInfo{
+		URLs: []string{
+			"brokenurl",
+			"https://example.com",
+		},
+	}
+
+	reader, _, err := imageSource.GetBlob(layerInfo)
+	require.NoError(t, err)
+	defer reader.Close()
+
+	data, err := ioutil.ReadAll(reader)
+	require.NoError(t, err)
+	assert.Contains(t, string(data), "Example Domain")
+}
+
+func TestGetBlobForRemoteLayersWithTLS(t *testing.T) {
+	imageSource := createImageSource(t, &types.SystemContext{
+		OCICertPath: "fixtures/accepted_certs",
+	})
+
+	layer, size, err := imageSource.GetBlob(types.BlobInfo{
+		URLs: []string{"https://localhost:12000/"},
+	})
+	require.NoError(t, err)
+
+	layerContent, _ := ioutil.ReadAll(layer)
+	assert.Equal(t, RemoteLayerContent, string(layerContent))
+	assert.Equal(t, int64(len(RemoteLayerContent)), size)
+}
+
+func TestGetBlobForRemoteLayersOnTLSFailure(t *testing.T) {
+	imageSource := createImageSource(t, &types.SystemContext{
+		OCICertPath: "fixtures/rejected_certs",
+	})
+	layer, size, err := imageSource.GetBlob(types.BlobInfo{
+		URLs: []string{"https://localhost:12000/"},
+	})
+
+	require.Error(t, err)
+	assert.Nil(t, layer)
+	assert.Equal(t, int64(0), size)
+}
+
+func remoteLayerContent(w http.ResponseWriter, req *http.Request) {
+	fmt.Fprintf(w, RemoteLayerContent)
+}
+
+func startRemoteLayerServer() (*http.Server, chan struct{}, error) {
+	certBytes, err := ioutil.ReadFile("fixtures/accepted_certs/cert.cert")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	clientCertPool := x509.NewCertPool()
+	if !clientCertPool.AppendCertsFromPEM(certBytes) {
+		return nil, nil, fmt.Errorf("Could not append certificate")
+	}
+
+	tlsConfig := &tls.Config{
+		// Reject any TLS certificate that cannot be validated
+		ClientAuth: tls.RequireAndVerifyClientCert,
+		// Ensure that we only use our "CA" to validate certificates
+		ClientCAs: clientCertPool,
+	}
+
+	tlsConfig.BuildNameToCertificate()
+
+	http.HandleFunc("/", remoteLayerContent)
+
+	httpServer := &http.Server{
+		Addr:      ":12000",
+		TLSConfig: tlsConfig,
+	}
+	httpServer.SetKeepAlivesEnabled(false)
+
+	serverStopSignal := make(chan struct{})
+	go func() {
+		httpServer.ListenAndServeTLS("fixtures/accepted_certs/cert.cert", "fixtures/accepted_certs/cert.key")
+		serverStopSignal <- struct{}{}
+	}()
+
+	return httpServer, serverStopSignal, nil
+}
+
+func createImageSource(t *testing.T, context *types.SystemContext) types.ImageSource {
+	imageSource, err := newImageSource(context, ociReference{
+		dir: "assets/manifest",
+	})
```

Comment:
(Non-blocking: Calling `NewReference` like a public user would make it obvious that there’s nothing special about constructing this reference.)

Reply:
So on this one it seems that go doesn't allow the `NewReference` return (`type.ImageReference`) to be used as an `ociReference`.

```
cannot use imageRef (type types.ImageReference) as type ociReference in argument to newImageSource: need type assertion
```

casting seems to work

Reply:
The public `ociReference.NewImageSource` can be called instead.